src/backend/distributed/operations/worker_shard_copy.c

```diff
@@ -0,0 +1,492 @@
+				copyDest->destinationShardFullyQualifiedName);
+			char *destinationShardRelationName = lsecond(
+				copyDest->destinationShardFullyQualifiedName);
+
+			ereport(ERROR, (errcode(ERRCODE_IO_ERROR),
+							errmsg("Failed to COPY to destination shard %s.%s",
+								   destinationShardSchemaName,
+								   destinationShardRelationName),
+							errdetail("failed to send %d bytes %s on node %u",
+									  copyDest->copyOutState->fe_msgbuf->len,
+									  copyDest->copyOutState->fe_msgbuf->data,
+									  copyDest->destinationNodeId)));
+		}
+
+		/* check whether there were any COPY errors */
+		PGresult *result = GetRemoteCommandResult(copyDest->connection,
+												  true /* raiseInterrupts */);
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			ReportCopyError(copyDest->connection, result);
+		}
+
+		PQclear(result);
+		ForgetResults(copyDest->connection);
+		CloseConnection(copyDest->connection);
+	}
+}
+
+
+/*
+ * CreateShardCopyDestReceiver creates a DestReceiver that copies into
+ * a destinationShardFullyQualifiedName on destinationNodeId.
+ */
+DestReceiver *
+CreateShardCopyDestReceiver(EState *executorState,
+							List *destinationShardFullyQualifiedName,
+							uint32_t destinationNodeId)
+{
+	ShardCopyDestReceiver *copyDest = (ShardCopyDestReceiver *) palloc0(
+		sizeof(ShardCopyDestReceiver));
+
+	/* set up the DestReceiver function pointers */
+	copyDest->pub.receiveSlot = ShardCopyDestReceiverReceive;
+	copyDest->pub.rStartup = ShardCopyDestReceiverStartup;
+	copyDest->pub.rShutdown = ShardCopyDestReceiverShutdown;
+	copyDest->pub.rDestroy = ShardCopyDestReceiverDestroy;
+	copyDest->pub.mydest = DestCopyOut;
+	copyDest->executorState = executorState;
+
+	copyDest->destinationNodeId = destinationNodeId;
+	copyDest->destinationShardFullyQualifiedName = destinationShardFullyQualifiedName;
+	copyDest->tuplesSent = 0;
+	copyDest->connection = NULL;
+	copyDest->useLocalCopy = CanUseLocalCopy(destinationNodeId);
+
+	return (DestReceiver *) copyDest;
+}
+
+
+/*
+ * ShardCopyDestReceiverDestroy frees the DestReceiver.
+ */
+static void
+ShardCopyDestReceiverDestroy(DestReceiver *dest)
+{
+	ShardCopyDestReceiver *copyDest = (ShardCopyDestReceiver *) dest;
+
+	if (copyDest->copyOutState)
+	{
+		pfree(copyDest->copyOutState);
+	}
+
+	if (copyDest->columnOutputFunctions)
+	{
+		pfree(copyDest->columnOutputFunctions);
+	}
+
+	pfree(copyDest);
+}
+
+
+/*
+ * ConstructCopyStatement constructs the text of a COPY statement
+ * for copying into a result table
+ */
+static StringInfo
+ConstructCopyStatement(List *destinationShardFullyQualifiedName, bool useBinaryFormat)
+{
+	char *destinationShardSchemaName = linitial(destinationShardFullyQualifiedName);
+	char *destinationShardRelationName = lsecond(destinationShardFullyQualifiedName);
+
+	StringInfo command = makeStringInfo();
+	appendStringInfo(command, "COPY %s.%s FROM STDIN",
+					 destinationShardSchemaName, destinationShardRelationName);
```

Comment:
are these quoted?

Reply:
No, I will wrap them in quote_identifier() call.